Домашняя работа №3

MAMKUH6OPIIIUK opened the pull request:
Относительно второй домашней работы:

1. Сборка/зависимости
- почти все зависимости теперь берутся из spring-boot-starter
- все зависимости для unit тестов берутся из spring-boot-starter-test
- изменен плагин сборки - теперь собирается fat jar

2. Запуск логики приложения теперь идет через бин, реализующий CommandLineRunner
3. Свойства перенесены в yaml файл и теперь грузятся при помощи ConfigurationProperties в тематические бины (для дао, для сервиса тестирования, для сервиса локализации). Соответственно, теперь в бины приложения все параметры инжектятся не через Value, а через передачу бина с загруженными в поля properties
4. Добавлен баннер с ANSI-цветами
5. Добавлены resource bundle для локализации на русский и английский языки
6. Добавлен сервис локализации (TranslationServiceImpl), в который инжектится MessageSource из автоконфигурации
7. Переработан сервис тестирования - теперь перед отправкой своих текстовок на IO запрашивает локализованный текст у сервиса локализации. Локализацию самих вопросов\ответов не производит.
8. Переработан бин дао. Задумка локализации вопросов и ответов в том, что csv файл остается в ресурсах, но теперь вместо фактических текстовых значений для вопросов и ответов хранит в себе ключи, по которым может быть запрошено фактическое текстовое значение из resource bundle при помощи MessageSource. Т.о. метод findAll() теперь грузит csv-конфиг и для каждой строки запрашивает локализованные значения нужных столбцов через сервис локализации. Т.е. на выходе сразу получается список локализованных вопросов. 

Тесты:

- теперь ни один из тестов не поднимает контекст Spring, т.к. в чате велели не поднимать. 

- добавлены моки сервиса локализации

MAMKUH6OPIIIUK committed:
Реализовал ДЗ №3

MAMKUH6OPIIIUK committed:
Добавил README для ДЗ №3

vlad-iv commented on spring-hw-03/.mvn/wrapper/maven-wrapper.properties:
```diff
@@ -0,0 +1,18 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Предлагаю все mvn* , .mvn, maven-* файлы 
положить в корень проекта (сразу в ветку main)
или удалить

MAMKUH6OPIIIUK replied:
Готово, удалил

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/config/ApplicationConfig.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@EnableConfigurationProperties({TestingDaoProperties.class, TestingServiceProperties.class,
```
Предлагаю объединить все *Properties классы в один
добавить, что он реализует 3 интерфейса (общепринято их называть "Имя"Provider , которые добавлять в зависимости в бины) 
Например interface LocaleProvider { Locale getLocale();}

MAMKUH6OPIIIUK replied:
Готово, объединил в один класс, реализующий несколько интерфейсов. Но из-за желания сделать некую иерархию свойств приложения в yaml-файле теперь приходится светить публичные setter'ы, принимающие Map'ы :(

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/config/TestingDaoProperties.java:
```diff
@@ -0,0 +1,16 @@
+
+@ConfigurationProperties(prefix = "testing.dao")
+public class TestingDaoProperties {
+    private String csv;
```
Предлагаю переименовать в String fileName или Resource resource
так как формат может изменить, а то что это ресурс нет

MAMKUH6OPIIIUK replied:
Готово, переименовал свойство в resource-basename в yaml.

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/domain/Question.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.Objects;
+
+public class Question {
+    private final String questionText;
```
из имен полей можно убрать имя сущности
questionText -> text ...

MAMKUH6OPIIIUK replied:
Готово. Убрал упоминания имен сущностей из всех полей

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/domain/Student.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Objects;
+
+public class Student {
+    private final String name;
```
Предлагаю попробовать Lombok 
@Getter, AllArgConstructor

MAMKUH6OPIIIUK replied:
👍 Готово, приделал везде и для всего, для чего можно

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/domain/Student.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Если функционал не требуется, можно его удалить ( equals и hashCode)

MAMKUH6OPIIIUK replied:
Готово, в классе Student убрал, во всех остальных классах модели заменил на аннотацию из Lombok (в них они нужны для тестов)

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/service/TranslationService.java:
```diff
@@ -0,0 +1,7 @@
+package ru.otus.spring.homework.oke.service;
+
+public interface TranslationService {
```
Общепринято: название сервис локализации Localize.....

MAMKUH6OPIIIUK replied:
Готово, переименовал интерфейс и реализацию

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/service/TranslationService.java:
```diff
@@ -0,0 +1,7 @@
+package ru.otus.spring.homework.oke.service;
+
+public interface TranslationService {
+    public String getTranslatedString(String code);
```
Тоже общепринятно getMessage

MAMKUH6OPIIIUK replied:
Готово

vlad-iv commented on spring-hw-03/src/main/resources/questions.csv:
```diff
@@ -0,0 +1,17 @@
+testing.theme1.name,select_one,testing.theme1.name.question1.text,testing.theme1.name.question1.answer1,3
```
Предлагаю создать для каждой локали свой csv файл
Так удобнее их поддерживать и добавлять новые

MAMKUH6OPIIIUK replied:
Готово.
Теперь в csv не лежат коды для поиска сообщений (из properties для MessageSource эти свойства удалены тоже).
Вместо этого есть несколько csv файлов с общим basename, но разными суффиксами локали в именах и с разным текстовым наполнением в зависимости от языка. 
Конкретный csv-ресурс выбирается сервисом локализации по запросу от дао (возвращается URL). Метод findAll в дао по-прежнему открывает inputStream к ресурсу при каждом обращении.

MAMKUH6OPIIIUK replied:
UPD: после занятия №8 поменял подход. 
Дао больше не зависит от сервиса локализации - сам ищет имя ресурса с учетом локали. Сообщения в исключениях теперь тоже не локализуются.
Соответственно, сервис локализации теперь занимается только локализацией сообщений и ничем больше

MAMKUH6OPIIIUK committed:
Исправления по результатам первой проверки ДЗ.
- Теперь в ресурсах хранится несколько csv-файлов с тестированиями под разные локали. URL для загрузки csv под сконфигурированную локаль ищется при помощи сервиса локализации (добавлен соответствующий метод)
- Удалены все mvn* файлы, добавлено их исключение в gitignore
- Все properties из yaml теперь грузятся через общий класс, реализующий несколько интерфейсов-провайдеров
- Изменена структура yaml файла, переименованы некоторые свойства, добавлены новые свойства (кодировка csv ресурса)
- В полях классов доменных моделей (и, соответственно, в getter) теперь не упоминаются имена сущностей, удален ненужный функционал
- Везде, где возможно, теперь используется Lombok для генерации конструкторов, getter и equals+hashcode
- Переименован сервис локализации и его методы

MAMKUH6OPIIIUK committed:
Поправил gitignore
Сделал покрасивее имя метода для доступа к свойству resource-basename

MAMKUH6OPIIIUK committed:
Перенес определение csv-ресурса для нужной локали из конструктора дао в его метод findAll

MAMKUH6OPIIIUK committed:
Исправления по результатам занятия №8 (что заметил у себя).
1. Сервис тестирования больше не возвращает неиспользуемый boolean. Вывод результатов тестирования пользователю - сам по себе результат работы
2. Сервис локализации больше не занимается поиском локализованных csv-ресурсов. Только обработка сообщений приложения
3. Дао теперь не зависит от сервиса локализации, чтобы не нарушать многослойную архитектуру. Т.е. сам ищет локализованный csv-ресурс по базовому имени и локали приложения. Не локализует сообщения в исключениях.
4. Имя реализации дао теперь отражает конкретную реализацию - поиск данных из локализованных csv
5. Имя кастомного исключения теперь не раскрывает конкретную реализацию дао

vlad-iv commented on spring-hw-03/.gitignore:
```diff
@@ -0,0 +1,36 @@
+HELP.md
```
Можно перенести в корень репозитория

MAMKUH6OPIIIUK replied:
Готово

vlad-iv commented on spring-hw-03/pom.xml:
```diff
@@ -0,0 +1,106 @@
+		<java.version>17</java.version>
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+		<opencsv.version>4.1</opencsv.version>
+		<lombok.version>1.18.26</lombok.version>
```
Можно удалить, версия берется из spring-boot-starter-parent

MAMKUH6OPIIIUK replied:
Готово

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/config/ApplicationPropertiesProvider.java:
```diff
@@ -0,0 +1,52 @@
+
+    private Locale locale;
+
+    private Map<String, Object> dao;
```
Предлагаю упростить, сразу инжектить значения в нужные поля (без map)
Для nested значений предпочтительнее создать отдельный класс или не использовать их.
https://www.baeldung.com/configuration-properties-in-spring-boot

MAMKUH6OPIIIUK replied:
Готово, убрал иерархию свойств, все свойства приложения теперь имеют только префикс testing и инжектятся сразу в нужные поля ApplicationPropertiesProvider. 
А как всё-таки правильно сделать конфигурацию с псевдоиерархией, чтобы хорошо было? Правильно ли я понимаю, что необходимо выполнить следующие действия?
Допустим, я хочу сделать что-то такое:
```
application:
	dao:
		resource: "azaza.txt"
		encoding: "UTF-8"
	service:
		some-property: "value"
```
Тогда я должен:
1. Сделать что-то вроде модели структур свойств. Т.е. класс DaoProperties с 2 полями - resource, encoding и класс ServiceProperties с одним полем - someProperty
2. Сделать 2 интерфейса: DaoPropertiesProvider, ServicePropertiesProvider, каждый из которых определяет методы для получения конкретных свойств
3. Сделать общий провайдер, например, ApplicationPropertiesProvider, который:
- имеет аннотацию `@ConfigurationProperties(prefix = "application")`
- имеет поля dao и service с типами DaoProperties, ServiceProperties, в которые будут инжектиться вложенные свойства
- реализует интерфейсы из п.2, в переопределяемых методах дергает геттеры у объектов своих полей dao и service.

vlad-iv replied:
можно сделать плоскую структуру.
да или иерархическую.
все верно только объекты для иерархической структуры хранят данные без реализации интерфейса

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoLocalizedCsv.java:
```diff
@@ -0,0 +1,250 @@
+ */
+@Repository
+public class TestingDaoLocalizedCsv implements TestingDao {
+    private static final String RESOURCE_EXTENSION = "csv";
```
Можно удалить, предлагаю имя файла с ресурсами получать из TestingDaoPropertiesProvider 
Так же предлагаю убрать любую логику (вычисление имени файла, локализация ...) из слоя Dao,
его ответственность только прочитать ресурс и вернуть список Testing

MAMKUH6OPIIIUK replied:
Готово, теперь, как и в дз №2, просто читает ресурс по его готовому имени, но само имя берется из провайдера

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoLocalizedCsv.java:
```diff
@@ -0,0 +1,250 @@
+     *
+     * @return URL найденного ресурса для установленной локали, либо null
+     */
+    private URL getResourceUrl() {
```
Можно удалить

MAMKUH6OPIIIUK replied:
Готово

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoLocalizedCsv.java:
```diff
@@ -0,0 +1,250 @@
+            throw new TestingLoadingException("CSV resource not found");
+        }
+        try (InputStream inputStream = localizedCsvResourceUrl.openStream()) {
+            readFromInputStream(inputStream, result);
```
Предлагаю изменить сигнатуру
List<Testing> result = readFromInputStream(inputStream);
Также можно переименовать метод, например в load

MAMKUH6OPIIIUK replied:
Готово, изменил сигнатуру и переименовал

vlad-iv commented on spring-hw-03/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoLocalizedCsv.java:
```diff
@@ -0,0 +1,250 @@
+    @Override
+    public List<Testing> findAll() {
+        List<Testing> result = new ArrayList<>();
+        URL localizedCsvResourceUrl = getResourceUrl();
```
Можно перенести в Provider или отказаться от логики, сразу передавать Resource или fileName

MAMKUH6OPIIIUK replied:
Перенёс определение полного имени ресурса по локали в yaml - в маску имени ресурса просто подставляется свойство testing.locale

MAMKUH6OPIIIUK committed:
Исправления по результатам второй итерации проверки дз №3.
1. Убрана иерархия в yaml. Все свойства приложения под одним префиксом.
2. Провайдер свойств
   2.1. Каждому свойству приложения свое поле, без Map. Отдельные классы под вложенные структуры свойств (как и сами вложенные структуры в yaml) создавать не стал, т.к. они уже были, но было предложено их заменить на один общий, реализующий несколько интерфейсов
   2.2. Вместо базового имени ресурса с вопросами теперь полное имя ресурса. Конструируется в самом yaml на основе значения свойства testing.locale.
3. Дао:
   3.1. Убрана логика поиска локализованного имени ресурса. Оно берется из соответствующего провайдера сразу готовое
   3.2. Изменена сигнатура и имя метода readFromInputStream -> loadFromInputStream. Теперь возвращает список тестирований.
   3.3. Переименован обратно в TestingDaoCsv, т.к. просто грузит данные из csv, не зная ничего о локали.
4. pom. Поднята версия, убрано переопределение уже заданных в spring-boot-parent свойство (версия java, кодировка исходного кода, версии lombok и enforce плагина)
5. .gitignore теперь только в корне репозитория

MAMKUH6OPIIIUK committed:
Убрал лишний getter из провайдера, т.к. он и так берется из lombok
